File: crates/armadra-hook/src/hook.rs
//! Hook mode: `armadra-hook <agentId>` with the CLI's hook payload on stdin.
//!
//! Hook mode is on the hot path of every event an agent CLI emits, so it is
//! written to be boring: it never fails loudly, never blocks longer than its
//! budget, and never writes to stdout except for a permission decision.

use std::fs;
use std::io::{self, Read, Write};
use std::path::Path;
use std::process;
use std::sync::mpsc;
use std::thread;
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

use serde_json::{json, Value};

use crate::endpoint::{env_var, is_valid_node_id};
use crate::http::{self, Request};
use crate::{Session, HOOK_PROTOCOL_VERSION, MAX_PAYLOAD_BYTES};

/// Exactly the JSON Claude expects back when a hook answers a permission
/// request. Emitted as literal text so key order is guaranteed.
const ALLOW_DECISION: &str = r#"{"hookSpecificOutput":{"hookEventName":"PermissionRequest","decision":{"behavior":"allow"}}}"#;
const DENY_DECISION: &str = r#"{"hookSpecificOutput":{"hookEventName":"PermissionRequest","decision":{"behavior":"deny","message":"由 Armadra 拒绝"}}}"#;

/// How often the permission answer file is checked.
const POLL_INTERVAL: Duration = Duration::from_millis(500);
/// How long we wait for the background "answered" report before exiting.
const ANSWERED_REPORT_TIMEOUT: Duration = Duration::from_millis(1500);

/// Runs hook mode. Always returns 0 — the caller is an agent CLI and a canvas
/// problem must never surface as a hook failure.
pub fn run(agent_id: &str) -> i32 {
    // Gate: outside a canvas node this binary is a no-op that still has to
    // consume stdin so the CLI's write side does not see a broken pipe.
    let Some(node_id) = env_var("ARMADRA_NODE_ID") else {
        drain_stdin();
        return 0;
    };

    let (bytes, truncated) = read_stdin_capped();
    let payload = build_payload(&bytes, truncated);

    let session = match Session::load() {
        Ok(session) => session,
        Err(error) => {
            debug(&error);
            return 0;
        }
    };

    if let Some(seconds) = permission_wait_secs(agent_id, &payload) {
        return run_permission_wait(&session, agent_id, payload, seconds);
    }

    let body = hook_body(&node_id, &payload, None, None);
    match post_hook(&session, agent_id, &body) {
        Ok(204) => {}
        Ok(status) => debug(&format!("hook endpoint answered {status}, expected 204")),
        Err(error) => debug(&error),
    }
    0
}

/// Builds the request body for `POST /hook/<agentId>`.
pub fn hook_body(
    node_id: &str,
    payload: &Value,
    pending_id: Option<&str>,
    answered: Option<&str>,
) -> Vec<u8> {
    let mut body = serde_json::Map::new();
    body.insert("nodeId".to_string(), json!(node_id));
    body.insert("version".to_string(), json!(HOOK_PROTOCOL_VERSION));
    body.insert("payload".to_string(), payload.clone());
    if let Some(pending_id) = pending_id {
        body.insert("pendingId".to_string(), json!(pending_id));
    }
    if let Some(answered) = answered {
        body.insert("answered".to_string(), json!(answered));
    }
    serde_json::to_vec(&Value::Object(body)).unwrap_or_else(|_| b"{}".to_vec())
}

fn post_hook(session: &Session, agent_id: &str, body: &[u8]) -> Result<u16, String> {
    let request = Request::post_json(
        format!("/hook/{}", percent_encode_segment(agent_id)),
        session.headers(),
        body.to_vec(),
    );
    http::send(&session.endpoint, &request).map(|response| response.status)
}

/// Reads stdin with a hard cap, reporting whether anything was dropped.
fn read_stdin_capped() -> (Vec<u8>, bool) {
    let stdin = io::stdin();
    let mut handle = stdin.lock();
    let mut buffer = Vec::new();
    // Read one byte past the cap so we can tell "exactly at the cap" from
    // "over the cap".
    let mut limited = (&mut handle).take(MAX_PAYLOAD_BYTES as u64 + 1);
    if limited.read_to_end(&mut buffer).is_err() {
        return (Vec::new(), false);
    }
    let truncated = buffer.len() > MAX_PAYLOAD_BYTES;
    if truncated {
        buffer.truncate(MAX_PAYLOAD_BYTES);
        // Keep consuming so the CLI writing the payload never sees EPIPE.
        let _ = io::copy(&mut handle, &mut io::sink());
    }
    (buffer, truncated)
}

fn drain_stdin() {
    let _ = io::copy(&mut io::stdin().lock(), &mut io::sink());
}

/// Turns raw stdin into the `payload` field.
///
/// Non-JSON input (or input we had to truncate, which cannot be valid JSON any
/// more) is wrapped as `{"raw": "..."}` so the runtime always sees an object.
pub fn build_payload(bytes: &[u8], truncated: bool) -> Value {
    let text = String::from_utf8_lossy(bytes);
    if truncated {
        return json!({ "raw": text, "truncated": true });
    }
    match serde_json::from_str::<Value>(text.trim()) {
        Ok(value) if value.is_object() || value.is_array() => value,
        _ => json!({ "raw": text }),
    }
}

/// Returns the wait budget when this invocation should answer a Claude
/// permission request in-hook, otherwise `None`.
pub fn permission_wait_secs(agent_id: &str, payload: &Value) -> Option<u32> {
    if agent_id != "claude" {
        return None;
    }
    if payload.get("hook_event_name").and_then(Value::as_str) != Some("PermissionRequest") {
        return None;
    }
    let seconds = env_var("ARMADRA_PERM_WAIT_SECS")?
        .trim()
        .parse::<u32>()
        .ok()?;
    if seconds == 0 {
        None
    } else {
        Some(seconds)
    }
}

/// Mints the id that ties a hook invocation to a canvas approval card.
///
/// Node id plus wall clock plus pid is unique enough: a single node cannot run
/// two hooks in the same millisecond from the same process.
pub fn pending_id(node_id: &str, epoch_ms: u128, pid: u32) -> String {
    format!("{node_id}-{epoch_ms}-{pid}")
}

fn now_epoch_ms() -> u128 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|delta| delta.as_millis())
        .unwrap_or(0)
}

fn run_permission_wait(session: &Session, agent_id: &str, payload: Value, seconds: u32) -> i32 {
    if !is_valid_node_id(&session.node_id) {
        debug("node id is not filesystem safe; skipping permission wait");
        let body = hook_body(&session.node_id, &payload, None, None);
        let _ = post_hook(session, agent_id, &body);
        return 0;
    }

    let pending_dir = session.endpoint.pending_dir();
    let pending_id = pending_id(&session.node_id, now_epoch_ms(), process::id());
    let request_path = pending_dir.join(format!("{pending_id}.json"));
    let answer_path = pending_dir.join(format!("{pending_id}.answer"));

    if let Err(error) = write_request_file(&pending_dir, &request_path, &payload) {
        // Without the request file the UI has nothing to render, so fall back
        // to a plain report and let the CLI show its own prompt.
        debug(&error);
        let body = hook_body(&session.node_id, &payload, None, None);
        let _ = post_hook(session, agent_id, &body);
        return 0;
    }

    let body = hook_body(&session.node_id, &payload, Some(&pending_id), None);
    if let Err(error) = post_hook(session, agent_id, &body) {
        debug(&error);
        let _ = fs::remove_file(&request_path);
        return 0;
    }

    match poll_for_answer(&answer_path, Duration::from_secs(u64::from(seconds))) {
        Some(decision) => {
            let _ = fs::remove_file(&answer_path);
            let _ = fs::remove_file(&request_path);
            // Report in the background: the decision is already made, so the
            // agent should not wait on the runtime to acknowledge it.
            let reported = report_answered(session, agent_id, &payload, &pending_id, decision);
            let mut stdout = io::stdout().lock();
            let _ = stdout.write_all(decision.output().as_bytes());
            let _ = stdout.write_all(b"\n");
            let _ = stdout.flush();
            drop(stdout);
            let _ = reported.recv_timeout(ANSWERED_REPORT_TIMEOUT);
            0
        }
        None => {
            // Fail open: print nothing and let the CLI fall back to its own
            // interactive prompt. The runtime sweeps orphaned request files.
            debug("permission wait timed out");
            0
        }
    }
}

/// The two decisions the canvas can hand back.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Decision {
    Allow,
    Deny,
}

impl Decision {
    pub fn parse(text: &str) -> Option<Decision> {
        match text.trim() {
            "allow" => Some(Decision::Allow),
            "deny" => Some(Decision::Deny),
            _ => None,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Decision::Allow => "allow",
            Decision::Deny => "deny",
        }
    }

    /// The exact stdout Claude parses.
    pub fn output(self) -> &'static str {
        match self {
            Decision::Allow => ALLOW_DECISION,
            Decision::Deny => DENY_DECISION,
        }
    }
}

fn poll_for_answer(path: &Path, budget: Duration) -> Option<Decision> {
    let deadline = Instant::now() + budget;
    loop {
        if let Ok(text) = fs::read_to_string(path) {
            if let Some(decision) = Decision::parse(&text) {
                return Some(decision);
            }
        }
        let left = deadline.saturating_duration_since(Instant::now());
        if left.is_zero() {
            return None;
        }
        thread::sleep(POLL_INTERVAL.min(left));
    }
}

/// Starts the "answered" report on a background thread.
///
/// The report is best effort: the decision has already been made locally, so a
/// slow runtime must not stall the agent. The returned receiver fires when the
/// POST finishes; the caller caps how long it is willing to wait.
fn report_answered(
    session: &Session,
    agent_id: &str,
    payload: &Value,
    pending_id: &str,
    decision: Decision,
) -> mpsc::Receiver<()> {
    let body = hook_body(
        &session.node_id,
        payload,
        Some(pending_id),
        Some(decision.as_str()),
    );
    let endpoint = session.endpoint.clone();
    let headers = session.headers();
    let path = format!("/hook/{}", percent_encode_segment(agent_id));
    let (sender, receiver) = mpsc::channel();
    thread::spawn(move || {
        let request = Request::post_json(path, headers, body);
        let _ = http::send(&endpoint, &request);
        let _ = sender.send(());
    });
    receiver
}

fn write_request_file(dir: &Path, path: &Path, payload: &Value) -> Result<(), String> {
    fs::create_dir_all(dir).map_err(|error| format!("cannot create {}: {error}", dir.display()))?;
    restrict_dir(dir);
    let bytes = serde_json::to_vec(payload).map_err(|error| error.to_string())?;
    let mut file = create_private(path)
        .map_err(|error| format!("cannot create {}: {error}", path.display()))?;
    file.write_all(&bytes)
        .map_err(|error| format!("cannot write {}: {error}", path.display()))?;
    file.flush().map_err(|error| error.to_string())
}

/// Creates a file only this user can read. On Windows the ACL inherited from
/// the per-user data directory is the best we can do without extra crates.
fn create_private(path: &Path) -> io::Result<fs::File> {
    let mut options = fs::OpenOptions::new();
    options.write(true).create(true).truncate(true);
    #[cfg(unix)]
    {
        use std::os::unix::fs::OpenOptionsExt;
        options.mode(0o600);
    }
    options.open(path)
}

fn restrict_dir(dir: &Path) {
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let _ = fs::set_permissions(dir, fs::Permissions::from_mode(0o700));
    }
    #[cfg(not(unix))]
    let _ = dir;
}

/// Percent-encodes a single path segment so an odd agent id cannot inject a
/// second path component or a query string.
pub fn percent_encode_segment(segment: &str) -> String {
    let mut out = String::with_capacity(segment.len());
    for byte in segment.bytes() {
        if byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'.' | b'~') {
            out.push(byte as char);
        } else {
            out.push_str(&format!("%{byte:02X}"));
        }
    }
    out
}

/// Diagnostics go to stderr and only when explicitly asked for; hook stdout is
/// reserved for the permission decision.
pub fn debug(message: &str) {
    if env_var("ARMADRA_HOOK_DEBUG").is_some() {
        let _ = writeln!(io::stderr(), "armadra-hook: {message}");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pending_id_shape() {
        assert_eq!(
            pending_id("node-7", 1_725_000_000_123, 4242),
            "node-7-1725000000123-4242"
        );
    }

    #[test]
    fn json_stdin_is_used_verbatim() {
        let payload = build_payload(br#"{"hook_event_name":"Stop"}"#, false);
        assert_eq!(payload["hook_event_name"], "Stop");
    }

    #[test]
    fn non_json_stdin_is_wrapped() {
        let payload = build_payload(b"not json at all", false);
        assert_eq!(payload["raw"], "not json at all");
        assert!(payload.get("truncated").is_none());
    }

    #[test]
    fn scalar_json_is_wrapped_too() {
        // A bare `12` is valid JSON but not an object; the runtime always wants
        // an object or array in `payload`.
        assert_eq!(build_payload(b"12", false)["raw"], "12");
    }

    #[test]
    fn oversize_stdin_is_marked_truncated() {
        let payload = build_payload(b"{\"a\":1}", true);
        assert_eq!(payload["truncated"], true);
        assert_eq!(payload["raw"], "{\"a\":1}");
    }

    #[test]
    fn payload_cap_matches_the_contract() {
        assert_eq!(MAX_PAYLOAD_BYTES, 1024 * 1024);
    }

    #[test]
    fn body_only_carries_optional_fields_when_set() {
        let payload = json!({"hook_event_name": "Stop"});
        let plain: Value = serde_json::from_slice(&hook_body("n1", &payload, None, None)).unwrap();
        assert_eq!(plain["nodeId"], "n1");
        assert_eq!(plain["version"], 1);
        assert!(plain.get("pendingId").is_none());
        assert!(plain.get("answered").is_none());

        let answered: Value =
            serde_json::from_slice(&hook_body("n1", &payload, Some("p1"), Some("deny"))).unwrap();
        assert_eq!(answered["pendingId"], "p1");
        assert_eq!(answered["answered"], "deny");
    }

    #[test]
    fn decisions_round_trip() {
        assert_eq!(Decision::parse(" allow\n"), Some(Decision::Allow));
        assert_eq!(Decision::parse("deny"), Some(Decision::Deny));
        assert_eq!(Decision::parse("maybe"), None);
        assert!(Decision::Allow.output().contains(r#""behavior":"allow""#));
        assert!(Decision::Deny.output().contains("由 Armadra 拒绝"));
        // Both outputs must be valid JSON for Claude to read them.
        serde_json::from_str::<Value>(Decision::Allow.output()).unwrap();
        serde_json::from_str::<Value>(Decision::Deny.output()).unwrap();
    }

    #[test]
    fn permission_mode_needs_all_three_conditions() {
        let request = json!({"hook_event_name": "PermissionRequest"});
        let stop = json!({"hook_event_name": "Stop"});
        // The env var is process wide, so exercise the non-env conditions only.
        assert!(permission_wait_secs("codex", &request).is_none());
        assert!(permission_wait_secs("claude", &stop).is_none());
    }

    #[test]
    fn agent_id_cannot_escape_its_path_segment() {
        assert_eq!(percent_encode_segment("claude"), "claude");
        assert_eq!(percent_encode_segment("custom:1"), "custom%3A1");
        assert_eq!(percent_encode_segment("../admin"), "..%2Fadmin");
    }

    #[test]
    fn poll_returns_none_without_an_answer() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("missing.answer");
        assert_eq!(poll_for_answer(&path, Duration::from_millis(10)), None);
    }

    #[test]
    fn poll_reads_the_answer_file() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("x.answer");
        fs::write(&path, "allow\n").unwrap();
        assert_eq!(
            poll_for_answer(&path, Duration::from_millis(10)),
            Some(Decision::Allow)
        );
    }

    #[test]
    fn request_files_are_private() {
        let dir = tempfile::tempdir().unwrap();
        let pending = dir.path().join("pending");
        let path = pending.join("n1-1-2.json");
        write_request_file(&pending, &path, &json!({"a": 1})).unwrap();
        assert_eq!(fs::read_to_string(&path).unwrap(), r#"{"a":1}"#);
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let mode = fs::metadata(&path).unwrap().permissions().mode();
            assert_eq!(mode & 0o777, 0o600);
            let dir_mode = fs::metadata(&pending).unwrap().permissions().mode();
            assert_eq!(dir_mode & 0o777, 0o700);
        }
    }
}

File: apps/desktop/src/cli/armadra-hook/hook.ts
/**
 * Hook mode: `armadra-hook <agentId>` with the CLI's hook payload on stdin.
 *
 * Hook mode is on the hot path of every event an agent CLI emits, so it is
 * written to be boring: it never fails loudly, never blocks longer than its
 * budget, and never writes to stdout except for a permission decision.
 */

import * as fs from "node:fs";
import * as path from "node:path";

import { envVar, isValidNodeId, pendingDir } from "./endpoint.js";
import type { Endpoint } from "./endpoint.js";
import { loadBinding } from "./context-usage.js";
import { canonicalJsonBytes, parseJson } from "./json.js";
import type { JsonValue } from "./json.js";
import { postJsonRequest, send as httpSend } from "./http.js";
import { headersFor, loadSession, send } from "./session.js";
import type { Session } from "./session.js";
import { HOOK_PROTOCOL_VERSION, MAX_PAYLOAD_BYTES } from "./usage.js";

/**
 * Exactly the JSON Claude expects back when a hook answers a permission
 * request. Emitted as literal text so key order is guaranteed.
 */
const ALLOW_DECISION =
  '{"hookSpecificOutput":{"hookEventName":"PermissionRequest","decision":{"behavior":"allow"}}}';
const DENY_DECISION =
  '{"hookSpecificOutput":{"hookEventName":"PermissionRequest","decision":{"behavior":"deny","message":"由 Armadra 拒绝"}}}';

/** How often the permission answer file is checked. */
const POLL_INTERVAL_MS = 500;
/** How long we wait for the background "answered" report before exiting. */
const ANSWERED_REPORT_TIMEOUT_MS = 1500;

export type Decision = "allow" | "deny";

export function parseDecision(text: string): Decision | undefined {
  const trimmed = text.trim();
  return trimmed === "allow" || trimmed === "deny" ? trimmed : undefined;
}

/** The exact stdout Claude parses. */
export function decisionOutput(decision: Decision): string {
  return decision === "allow" ? ALLOW_DECISION : DENY_DECISION;
}

/**
 * Runs hook mode. Always returns 0 — the caller is an agent CLI and a canvas
 * problem must never surface as a hook failure.
 */
export async function run(agentId: string): Promise<number> {
  // Gate: outside a canvas node this client is a no-op that still has to
  // consume stdin so the CLI's write side does not see a broken pipe.
  const nodeId = envVar("ARMADRA_NODE_ID");
  if (nodeId === undefined) {
    await drainStdin();
    return 0;
  }

  const binding = loadBinding();
  const { bytes, truncated } = await readStdinCapped();
  const payload = buildPayload(bytes, truncated);

  let session: Session;
  let terminalBinding: JsonValue | undefined;
  if (binding !== undefined) {
    session = binding.session;
    terminalBinding = {
      sessionId: binding.sessionId,
      generation: binding.generation,
      sourceRevision: String(binding.revision),
    };
  } else {
    const loaded = loadSession();
    if ("error" in loaded) {
      debug(loaded.error);
      return 0;
    }
    session = loaded.ok;
  }

  const seconds = permissionWaitSecs(agentId, payload);
  if (seconds !== undefined)
    return runPermissionWait(session, agentId, payload, seconds);

  const body = hookBody(nodeId, payload, undefined, undefined, terminalBinding);
  const outcome = await postHook(session, agentId, body);
  if ("error" in outcome) debug(outcome.error);
  else if (outcome.status !== 204) {
    debug(`hook endpoint answered ${outcome.status}, expected 204`);
  }
  return 0;
}

/** Builds the request body for `POST /hook/<agentId>`. */
export function hookBody(
  nodeId: string,
  payload: JsonValue,
  pendingId?: string,
  answered?: string,
  terminalBinding?: JsonValue,
): Buffer {
  return canonicalJsonBytes({
    nodeId,
    version: HOOK_PROTOCOL_VERSION,
    payload,
    pendingId,
    answered,
    terminalBinding,
  });
}

async function postHook(
  session: Session,
  agentId: string,
  body: Buffer,
): Promise<{ status: number } | { error: string }> {
  const route = `/hook/${percentEncodeSegment(agentId)}`;
  const outcome = await send(session, (current, candidate) =>
    postJsonRequest(route, headersFor(current, candidate), body),
  );
  return "error" in outcome ? outcome : { status: outcome.ok.status };
}

/** Reads stdin with a hard cap, reporting whether anything was dropped. */
export async function readStdinCapped(): Promise<{
  bytes: Buffer;
  truncated: boolean;
}> {
  const chunks: Buffer[] = [];
  let kept = 0;
  let total = 0;
  try {
    for await (const chunk of process.stdin) {
      const buffer = chunk as Buffer;
      total += buffer.length;
      // Keep one byte past the cap so we can tell "exactly at the cap" from
      // "over the cap", but keep consuming either way so the CLI writing the
      // payload never sees EPIPE.
      if (kept <= MAX_PAYLOAD_BYTES) {
        const room = MAX_PAYLOAD_BYTES + 1 - kept;
        const slice = buffer.subarray(0, room);
        chunks.push(slice);
        kept += slice.length;
      }
    }
  } catch {
    return { bytes: Buffer.alloc(0), truncated: false };
  }
  const truncated = total > MAX_PAYLOAD_BYTES;
  const bytes = Buffer.concat(chunks);
  return {
    bytes: truncated ? bytes.subarray(0, MAX_PAYLOAD_BYTES) : bytes,
    truncated,
  };
}

async function drainStdin(): Promise<void> {
  try {
    for await (const _chunk of process.stdin) {
      // Discarded on purpose: the write side must not see a broken pipe.
    }
  } catch {
    // A closed stdin is the normal case outside a hook.
  }
}

/**
 * Turns raw stdin into the `payload` field.
 *
 * Non-JSON input (or input we had to truncate, which cannot be valid JSON any
 * more) is wrapped as `{"raw": "..."}` so the runtime always sees an object.
 */
export function buildPayload(bytes: Buffer, truncated: boolean): JsonValue {
  const text = bytes.toString("utf8");
  if (truncated) return { raw: text, truncated: true };
  try {
    const value = parseJson(text.trim());
    if (value !== null && typeof value === "object") return value;
  } catch {
    // Falls through to the raw wrapper below.
  }
  return { raw: text };
}

/**
 * Returns the wait budget when this invocation should answer a Claude
 * permission request in-hook, otherwise `undefined`.
 */
export function permissionWaitSecs(
  agentId: string,
  payload: JsonValue,
): number | undefined {
  if (agentId !== "claude") return undefined;
  if (payload === null || typeof payload !== "object" || Array.isArray(payload))
    return undefined;
  if (
    (payload as Record<string, JsonValue>)["hook_event_name"] !==
    "PermissionRequest"
  ) {
    return undefined;
  }
  const raw = envVar("ARMADRA_PERM_WAIT_SECS");
  if (raw === undefined) return undefined;
  const text = raw.trim();
  if (!/^\d+$/.test(text)) return undefined;
  const seconds = Number(text);
  return seconds === 0 || seconds > 0xffff_ffff ? undefined : seconds;
}

/**
 * Mints the id that ties a hook invocation to a canvas approval card.
 *
 * Node id plus wall clock plus pid is unique enough: a single node cannot run
 * two hooks in the same millisecond from the same process.
 */
export function pendingId(
  nodeId: string,
  epochMs: number,
  pid: number,
): string {
  return `${nodeId}-${epochMs}-${pid}`;
}

/**
 * Unlike {@link postHook}, this cannot hand the failover loop a single
 * candidate-agnostic closure: the pending request/answer files have to sit
 * next to whichever candidate's endpoint file actually accepts the report,
 * because that is the directory the *answering* runtime watches. So this walks
 * the candidates itself, writing the pending file fresh for each one it tries
 * and removing it the moment that candidate turns out not to be the live one —
 * only a transport failure moves on to the next candidate; an HTTP-level
 * rejection is authoritative and falls back to a plain (non-waiting) report on
 * the spot (§2.5).
 */
async function runPermissionWait(
  session: Session,
  agentId: string,
  payload: JsonValue,
  seconds: number,
): Promise<number> {
  if (!isValidNodeId(session.nodeId)) {
    debug("node id is not filesystem safe; skipping permission wait");
    await postHook(session, agentId, hookBody(session.nodeId, payload));
    return 0;
  }

  const id = pendingId(session.nodeId, Date.now(), process.pid);
  const route = `/hook/${percentEncodeSegment(agentId)}`;
  let lastError = "";

  for (const candidate of session.candidates) {
    const directory = pendingDir(candidate);
    const requestPath = path.join(directory, `${id}.json`);
    const answerPath = path.join(directory, `${id}.answer`);

    const written = writeRequestFile(directory, requestPath, payload);
    if (written !== undefined) {
      debug(written);
      lastError = written;
      continue;
    }

    const body = hookBody(session.nodeId, payload, id);
    const outcome = await httpSend(
      candidate,
      postJsonRequest(route, headersFor(session, candidate), body),
    );
    if ("ok" in outcome && outcome.ok.status === 204) {
      return awaitDecision(
        session,
        candidate,
        agentId,
        payload,
        id,
        requestPath,
        answerPath,
        seconds,
      );
    }
    if ("ok" in outcome) {
      // Reached the runtime and it answered — authoritative, per §2.5. Fall
      // back to a plain report rather than trying another candidate that never
      // saw this pending id.
      removeQuietly(requestPath);
      debug("permission report was rejected by the endpoint");
      await postHook(session, agentId, hookBody(session.nodeId, payload));
      return 0;
    }
    // Transport failure: nothing is listening here, so nothing is watching the
    // pending file we just wrote either.
    removeQuietly(requestPath);
    lastError = outcome.error;
  }
  // Fail open: print nothing and let the CLI fall back to its own interactive
  // prompt. The runtime sweeps orphaned request files.
  debug(lastError);
  return 0;
}

async function awaitDecision(
  session: Session,
  candidate: Endpoint,
  agentId: string,
  payload: JsonValue,
  id: string,
  requestPath: string,
  answerPath: string,
  seconds: number,
): Promise<number> {
  const decision = await pollForAnswer(answerPath, seconds * 1000);
  if (decision === undefined) {
    debug("permission wait timed out");
    return 0;
  }
  removeQuietly(answerPath);
  removeQuietly(requestPath);
  // Report in the background: the decision is already made, so the agent
  // should not wait on the runtime to acknowledge it. Reported on the same
  // candidate the request went to — not a fresh failover search — because that
  // is the runtime that holds the pending record this answers.
  const reported = httpSend(
    candidate,
    postJsonRequest(
      `/hook/${percentEncodeSegment(agentId)}`,
      headersFor(session, candidate),
      hookBody(session.nodeId, payload, id, decision),
    ),
  );
  process.stdout.write(`${decisionOutput(decision)}\n`);
  await withTimeout(reported, ANSWERED_REPORT_TIMEOUT_MS);
  return 0;
}

/**
 * Waits for `work`, but no longer than `ms`. The timer is cleared as soon as
 * `work` settles, so a report that answers in 3 ms does not hold the process
 * open for the rest of the budget.
 */
function withTimeout(work: Promise<unknown>, ms: number): Promise<void> {
  return new Promise((resolve) => {
    const timer = setTimeout(resolve, ms);
    void work.then(
      () => {
        clearTimeout(timer);
        resolve();
      },
      () => {
        clearTimeout(timer);
        resolve();
      },
    );
  });
}

export async function pollForAnswer(
  file: string,
  budgetMs: number,
): Promise<Decision | undefined> {
  const deadline = Date.now() + budgetMs;
  for (;;) {
    try {
      const decision = parseDecision(fs.readFileSync(file, "utf8"));
      if (decision !== undefined) return decision;
    } catch {
      // Not answered yet.
    }
    const left = deadline - Date.now();
    if (left <= 0) return undefined;
    await delay(Math.min(POLL_INTERVAL_MS, left));
  }
}

/**
 * A referenced timer on purpose: the poll loop is the only thing keeping the
 * process alive between checks, and an unreferenced one lets Node decide there
 * is no work left and exit before the answer lands.
 */
function delay(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

function removeQuietly(file: string): void {
  try {
    fs.rmSync(file, { force: true });
  } catch {
    // Already gone.
  }
}

/** Writes the 0600 request file, answering with an error message on failure. */
export function writeRequestFile(
  directory: string,
  file: string,
  payload: JsonValue,
): string | undefined {
  try {
    fs.mkdirSync(directory, { recursive: true });
  } catch (error) {
    return `cannot create ${directory}: ${message(error)}`;
  }
  restrictDir(directory);
  try {
    // On Windows the ACL inherited from the per-user data directory is the
    // best we can do without extra dependencies.
    const handle = fs.openSync(file, "w", 0o600);
    try {
      fs.writeFileSync(handle, canonicalJsonBytes(payload));
    } finally {
      fs.closeSync(handle);
    }
  } catch (error) {
    return `cannot create ${file}: ${message(error)}`;
  }
  return undefined;
}

function restrictDir(directory: string): void {
  if (process.platform === "win32") return;
  try {
    fs.chmodSync(directory, 0o700);
  } catch {
    // Best effort, exactly as in the Rust client.
  }
}

function message(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

/**
 * Percent-encodes a single path segment so an odd agent id cannot inject a
 * second path component or a query string.
 */
export function percentEncodeSegment(segment: string): string {
  let out = "";
  for (const byte of Buffer.from(segment, "utf8")) {
    const character = String.fromCharCode(byte);
    if (/[A-Za-z0-9\-_.~]/.test(character)) out += character;
    else out += `%${byte.toString(16).toUpperCase().padStart(2, "0")}`;
  }
  return out;
}

/**
 * Diagnostics go to stderr and only when explicitly asked for; hook stdout is
 * reserved for the permission decision.
 */
export function debug(text: string): void {
  if (envVar("ARMADRA_HOOK_DEBUG") !== undefined) {
    process.stderr.write(`armadra-hook: ${text}\n`);
  }
}
